cmd/gardenlet/app/migration.go

```diff
@@ -377,3 +379,33 @@ func migrateDeprecatedTopologyLabels(ctx context.Context, log logr.Logger, seedC
 
 	return flow.Parallel(taskFns...)(ctx)
 }
+
+// TODO(plkokanov): Remove this code after gardener v1.96 has been released.
+// The refactoring done in https://github.com/gardener/gardener/pull/8061 incorrectly changed
+// the label used to select vali vpas by the corresponding vali hvpa from `role: vali-vpa` to
+// `role: valivpa`. This caused the hvpa controller to orphan the existing vpa objects with
+// the `role: vali-vpa` label and create new ones.
+// deleteOrphanedValiVPAs deletes the orphaned vali vpas with the `role vali-vpa` label.
+func deleteOrphanedValiVPAs(ctx context.Context, log logr.Logger, c client.Client) error {
+	var (
+		orphanedValiVPALabel = "vali-vpa"
+		orphanedVPAList      = &vpaautoscalingv1.VerticalPodAutoscalerList{}
+	)
+
+	if err := c.List(ctx, orphanedVPAList, client.MatchingLabels(map[string]string{v1beta1constants.LabelRole: orphanedValiVPALabel})); err != nil {
+		return fmt.Errorf("could not list orphaned vali VPAs with label '%s: %s': %w", v1beta1constants.LabelRole, orphanedValiVPALabel, err)
+	}
+
+	fns := make([]flow.TaskFn, 0, len(orphanedVPAList.Items))
+	for _, obj := range orphanedVPAList.Items {
+		fns = append(fns, func(ctx context.Context) error {
+			log.Info("Deleting orphaned vali VPA", "vpa", client.ObjectKeyFromObject(&obj))
+			if err := client.IgnoreNotFound(c.Delete(ctx, &obj)); err != nil {
```

Comment:
I guess `obj `this is a loop var and you have to make a copy of it. See the problem description in https://go.dev/blog/loopvar-preview.

Reply:
This should be fixed for go 1.22 so that you no longer require a copy, right?
Additionally, I did test it with multiple VPA objects. They all get deleted as expected.

Reply:
OK 👍